### ISSUE TITLE
Deprecate the "UberDOG" concept

### ISSUE THREAD

**Issue by CFSworks** [Astron/Astron#205]
I fear that the concept of an UberDOG (even if renamed to something less confusing like "static object") is too confusing for beginners. Additionally, I think it has some drawbacks:
- It breaks the "objects are part of a tree" rule. (As these objects cannot have parents or children.)
- It breaks the "objects can have state" rule.
- It requires redundant configuration in the UD server, the CA, _and_ the client.
- It requires static IDs to work well, which may become a problem if we decide to make the SS allocate IDs.

As "service objects" are useful, I do not recommend getting rid of that concept. However, I suggest they work differently: rather than sitting on a well-known channel, they should instead sit in a well-known location. The CA can be configured to open interest on this well-known location automatically upon HELLO, so that clients automatically discover the service objects (however, a different mechanism would be needed to specify anonymous privilege as there would no longer be an "uberdogs: / anonymous: true" field in the YAML file -- perhaps only anonymous objects are visible after HELLO and the LoginManager opens a server-side interest after authentication?)

For AI<->AI communication, we would need to revise the InternalRepository "specification" to support opening interest on locations as well. This allows AIs to become aware of each other's objects without needing explicit messaging channels between the two. We should also amend the API spec. to include a mechanism so that AI servers know what objects belong to them (i.e. have their AI channel set) and what objects are merely "seen" (i.e. the AI server has interest on the object's location). Bonus points if we add an "aicast" keyword, which is an AI-only broadcast (where updates only sent to AI servers with interest; clients do not receive them).

In order to phase out UberDOGs under this plan, we would need to:
- [ ] Deprecate the "uberdog:" section in the configuration file.
- [ ] Add a way to configure automatic interest in a Client Agent. Newly-connecting clients thus automatically open interest in the automatically-configured zones.
- [ ] Add a way to define well-known root objects in the visibility graph. Perhaps a special "parent object" role?
- [ ] Amend InternalRepository implementations to support opening interest, and keep track of which objects are AI-owned and which objects are merely seen via one or more open interests.

As this would have a pretty big impact, I would suggest not merging it in until we're ready to do a minor version bump.

**Comment by mfwass** [comment]
One of my concerns with the deprecation of the UberDOG system would be the already existing client authentication systems. Currently right now the only way to authenticate a client is through an UberDOG(as far as I know). In TTR's case, the ```ClientServicesManagerUD```, or CSMUD for short, handles authentication. Those systems would need to be rewritten if the UberDOG's are deprecated(inb4 legacy branch).

I think that Astron should aim for something similar to how Toontown's authentication worked. My idea for the implementation of a system like this would be a configuration for the CA with a script loaded by the CA that is wrote by the developer using Astron. This script would be the equivalent of the CSMUD. This way, you can configure your server to use authentication or to not use it, and if you choose to use authentication, you can write it however you want to allowing customization and possible better security.

Here is a proposed example for how the CA authentication could be configured in ```astrond.yml```:

No-authentication CA config (w/o TLS):

```
    - type: clientagent
      bind: 0.0.0.0:6667
      version: "FooGame v7.0"
      client:
        type: fooclient
      channels:
          min: 100100
          max: 100999
```
 
Authentication-enabled CA config (w/o TLS):
```
    - type: clientagent
      bind: 0.0.0.0:6667
      version: "FooGame v7.0"
      client:
        type: fooclient
      channels:
          min: 100100
          max: 100999
      authentication:
          filename: FooGameCSM.py
```

I have talked to @CFSworks about this before, and he said he likes the idea.

**Comment by xHarv14** [comment]
@mfwass The thing here is, it's not Astron's job to handle authentication (which becomes more applicatoin specific). With the deprecation of UberDOGs, we'd have "unauthenticated" fields for AIs. Any unauthenticated client can send an update to that to login (much like how they do so with UberDOGs currently).

With that being said, one of the plans is to modularize Astron, and any developer using Astron could easily create an authentication module and load it up in their daemon, that's entirely up to them.

TL;DR: We aren't _removing_ the ability to authenticate via game-specific code, we're just moving it.

**Comment by mfwass** [comment]
@xHarv14 Understood.
